### PR TITLE
Avoid overwriting closures while initialising recursive modules

### PR DESCRIPTION
**Pull request by stedolan**
This patch switches to a different mechanism for initialising recursive modules (which should avoid the bug in #10203) as well as adding some more testing of this mechanism. The new `recursive_module_init` test has been verified to give the same result before and after this patch.

The new mechanism never overwrites a closure. Instead, it overwrites the relevant slot in the module block. The stub closure created during initialisation always checks the relevant slot, so it doesn't need to be mutated. In most cases, I'd expect the stub closure to be garbage-collected soon after initialisation, so I expect this to have equivalent performance to the current system.

(The only cases where the stub will continue to be used after initialisation are those where a reference to a function is created during initialisation via the recursive module name, but outside of any lambdas. See `M.f3` in the test case. I don't think this will happen much in real code, and if it does pose a problem it can be worked around by eta-expanding)

### CODE REVIEW

**Comment by xavierleroy** [comment]
That's an interesting alternative to the in-place update of closures.  I just hope it's not breaking an assumption that blocks representing structures are immutable, in flambda or elsewhere.

> The only cases where the stub will continue to be used after initialisation are those where a reference to a function is created during initialisation via the recursive module name, but outside of any lambdas. See M.f3 in the test case. I don't think this will happen much in real code, and if it does pose a problem it can be worked around by eta-expanding

I'm thinking of code like the following:
```
module rec A ... = struct let f =  List.map B.g ... end
and B ... = struct let g x = ... end
```
In this case the stub `B.g` is captured by the partial application `List.map B.g`.  If I understand correctly, when `A.f` is applied, it will redirect to the final `B.g` function?  So it's a bit slower but still correct?

**Comment by stedolan** [comment]
> ```
> module rec A ... = struct let f =  List.map B.g ... end
> and B ... = struct let g x = ... end
> ```
> 
> In this case the stub `B.g` is captured by the partial application `List.map B.g`. If I understand correctly, when `A.f` is applied, it will redirect to the final `B.g` function? So it's a bit slower but still correct?

Yes, exactly. The stub is something like:
```ocaml
let rec g_stub x = if B.g == g_stub then raise Undefined ... else B.g x
```
If performance matters, the stub can be avoided by eta-expanding `f` to:
```ocaml
let f x = List.map B.g x
```
(This is probably a good idea for performance anyway, since List.map is a two-argument function and calling it with two arguments avoids the indirection through caml_curry)

**Comment by alainfrisch** [comment]
Just to confirm, in a code like:

````ocaml
module F(X : sig val y: unit->string end) = struct
  let f () = print_string (X.y ())
end

module rec A : sig val f: unit -> unit val y: unit->string
end = struct
  let y () = "HELLO"
  include F(A)
end

let () = A.f ()
````

the function call `X.y()` will go through the initial stub (with the dynamic check), right (because of module coercion, the structure for A cannot directly be passed to F), right?

**Comment by stedolan** [comment]
> the function call X.y() will go through the initial stub (with the dynamic check), right (because of module coercion, the structure for A cannot directly be passed to F), right?

Yes. (It's possible to work around this by expanding `F(A)` to `F(struct let y = y end)`, though)

Come to think of it, an alternative is possible here: functions that don't actually have any free occurrences of the recursive modules (e.g. your `y`) don't need backpatching, so we could just initialise those early.

**Comment by stedolan** [comment]
> Come to think of it, an alternative is possible here: functions that don't actually have any free occurrences of the recursive modules (e.g. your y) don't need backpatching, so we could just initialise those early.

To expand on this a little after some git-spelunking:

There's already a mechanism in camlinternalMod to allow some members of a module to be initialised eagerly and never patched. It was introduced by @garrigue in 2014 to fix a case involving module aliases (#6307) but not otherwise used. This use was removed by @lpw25 in #1610, so the `Value` case of `CamlinternalMod.shape` is now completely unused, but as far as I can tell could be used here easily enough.

**Comment by xavierleroy** [comment]
OK, I'm convinced the proposed approach gives the same results as the current implementation.

Performance-wise, here is my analysis:
- The proposed approach results in one call to an unknown function when the function is obtained through the module (`M.f`), and two such calls + an array access (`Obj.obj modu.(i)`) + a dynamic test (`if fn == fn'`) if the stub function value is used.
- The workaround I proposed in #10203 (always use the "trampoline" approach) costs two calls to unknown functions in both cases.

Note that we can have our cake and eat it too: overwrite the slot in the module block AND overwrite the closure with a trampoline that calls the final function.  Not sure the performance gain is worth the effort.

**Comment by hhugo** [comment]
I've used similar approach to implement recursive module in jsoo by replacing camlintermalMod entirely. I suspect that it would work out of the box if/once that PR get merged. I'll give it a try and report back.

**Comment by mshinwell** [comment]
@stedolan and I discussed whether this proposal was safe with Flambda or similar.  We believe it is, since the module blocks are allocated as arrays, which are expected to be mutable and whose contents are not tracked.  However we think the safety could be made more explicit by using `Obj.new_block` instead, which @stedolan is going to look at.

@lpw25 had a concern that this new approach might not be safe with some caching that happens in the object system runtime (I don't know the details) but I think he was convinced in the end that everything will be ok, thanks to the physical equality check, that allows one to safely capture the initial closures.

**Comment by stedolan** [comment]
I've switched from `Obj.t array` to `Obj.new_block` and friends.

**Comment by stedolan** [comment]
@xavierleroy What's the status of this? Are you happy with the current implementation or do you want changes?

File: stdlib/camlinternalMod.ml
Comment (xavierleroy) [inline stdlib/camlinternalMod.ml]: If this is no longer used, please remove the declaration, don't stick an underscore in front to silence the warning.

**Review by xavierleroy — COMMENTED** [review]
> @xavierleroy What's the status of this? Are you happy with the current implementation or do you want changes?

I'm globally in favor of this change, as it makes the code more robust and also simpler.  I haven't had time to review it line by line, though, so a second review and approval is needed. 

Also, the code needs cleaning up, see below.

File: stdlib/camlinternalMod.ml
Comment (hhugo) [inline stdlib/camlinternalMod.ml]: One could lift `Obj.field modu i` outside the loop.

**Comment by stedolan** [comment]
> Also, the code needs cleaning up, see below.

Done (and rebased onto trunk)

**Comment by stedolan** [comment]
I've just rebased this onto trunk again. Anyone have time to have a look? I'd like to get no-naked-pointers mode fixed.

**Comment by lthls** [comment]
I've been looking for observable changes caused by this PR, and I finally found one:
```
let r : (int -> int) ref = ref (fun x -> x)

module rec M : sig val f : (int -> int) Lazy.t end =
struct
  let _ =
    match Lazy.force M.f 0 with
    | _ -> ()
    | exception Undefined_recursive_module _ -> ()

  let _ =
    let f = M.f in
    r := (fun x -> Lazy.force f x)

  let f = lazy (fun x -> x + 1)
end

let _ =
  Format.printf "Normal version: %d@." (Lazy.force M.f 0);
  Format.printf "Escaped version: %d@." (!r 0);
  ()
```
With trunk, both versions print a result of `1`, while with this PR the escaped version will raise an `Undefined_recursive_module` exception.

As you can see, it's not very likely to occur in the wild. Besides, which result is correct is not clear (I think I prefer the result after this PR). So I think this should not prevent this PR from being merged.

Since I have already looked at most of the code of this PR in detail, I'll write a proper review, hopefully later today.

**Comment by stedolan** [comment]
That is an excellent example, well spotted.

I think I can still make the following two backwards-compatibility claims, though, which should cover any reasonable program:

  - The set of programs that raise `Undefined_recursive_module` has not ~~changed~~ grown *(see @hhugo's comment below)*.
  - The behaviour of those programs that never raise `Undefined_recursive_module` has not changed.

The behaviour of some programs that raise `Undefined_recursive_module` has changed, though: as your example shows, a lazy value that raised `Undefined_recursive_module` will now always raise that exception, while 4.12 may eventually overwrite it with a fresh unforced lazy value that does not raise. (However, to notice this one must catch Undefined_recursive_module and then inspect the half-formed recursive module)

**Comment by lpw25** [comment]
I think the old behaviour is the correct one. I think the intended behaviour is that attempting to force a lazy value too soon raises an exception. Whereas the new behaviour is something like "this lazy acts as if it was defined with a function that always raises until the point at which the lazy is defined".

Getting the old behaviour with the new approach seems like it would need some changes to the lazy machinery though, so its probably not worth it.

**Comment by stedolan** [comment]
The old behaviour means that there exist Lazy.t values which produce different results when forced twice. Perhaps this is the intended behaviour here, but I don't like the weirdness of recursive modules affecting the otherwise relatively unweird Lazy.t.

I would say that both the old and the new behaviour can be accurately described by "attempting to force a lazy value too soon raises an exception". They differ only in what later happens to that lazy value if it is in fact forced too soon: should further forcings of the same lazy value also raise?

This difference can only be observed by catching `Undefined_recursive_module`, and then inspecting the recursive module. I don't think it's important what happens here: the name of the exception is a good hint that the recursive module might not be well-defined.

**Comment by lpw25** [comment]
I don't think the new behaviour does match "attempting to force a lazy value too soon raises an exception". The distinction I am trying to get at is that, in the new behaviour, it is not the act of forcing that causes the exception. Instead, the behaviour of forcing is unchanged but we are forcing a function whose behaviour depends on when it is called.

I agree that fixing this corner case isn't worth the effort.

**Review by lthls — APPROVED** [review]
I've reviewed the code carefully and I believe it is correct.
Here are a few comments on what I found during my review:

- Observable changes are very limited. With the exception of the discussion about lazy values, the only way I could find to detect whether this patch has been applied or not is to use physical equality (or `Obj` functions). (Note that you don't need to inspect a recursive module being defined to notice the difference. Example:
```
let f x = x
module rec M : sig val g : int -> int end = struct
  let g = f
end
let b = M.g == f
```
Physical equality has always been allowed to return both true and false for allocated values that are semantically equal, so that's not a problem.)
- I'm not completely convinced that the code for handling classes is safe, but I'm convinced that this PR uses the exact same mechanism as before.
- The performance implications have already been summed up by @xavierleroy
- I've discussed the issue of mutating the field of an immutable block with @mshinwell and @stedolan, and we believe that it's fine. First, the fact that the block is expected to be immutable cannot be propagated at the moment (the use of `Obj.new_block` means that the middle-end and backend will not assume that the block is immutable). In Flambda 2, we have added mutability annotations on field reads though so we can end up in a situation where we're allowed to reuse the result of a field read occurring before the module update, instead of re-doing the field access. However, this would be both safe and very unlikely to happen. Safe because after the module field is set both values are semantically equivalent anyway (except for the corner case of already evaluated lazy values). And unlikely because Flambda 2 uses block approximations, not CSE, to simplify field accesses; the backend CSE might trigger here in theory, but in practice the calls to `update_mod` between the reads would almost certainly prevent the load equations from being propagated.

So this is good to merge as far as I'm concerned.

**Comment by hhugo** [comment]
> * The set of programs that raise `Undefined_recursive_module` has not changed.

The following program raises `Undefined_recursive_module` with 4.12 and 4.11 but runs successfully in 4.13
```ocaml
module rec M : sig
  val f : int -> int

  val g : int -> int
end = struct
  let f = M.g

  let g i = if i < 0 then -1 else M.f (i - 1)
end
;;
M.g 10
```

**Comment by stedolan** [comment]
Well spotted! I did not think of that.

I still think that any program which does not raise Undefined_recursive_module in 4.12 will have the same behaviour in 4.13 (modulo the physical equality differences noticed by @lthls). Programs that do raise in 4.12 might raise at a different time or, apparently, not raise any more!

**Pull request merged**
